src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringConcat.cs

```diff
@@ -346,6 +406,156 @@ private BoundExpression RewriteStringConcatenationManyExprs(SyntaxNode syntax, I
             return BoundCall.Synthesized(syntax, receiverOpt: null, initialBindingReceiverIsSubjectToCloning: ThreeState.Unknown, method, array);
         }
 
+        private bool TryRewriteStringConcatenationWithSpanBasedConcat(SyntaxNode syntax, ArrayBuilder<BoundExpression> args, [NotNullWhen(true)] out BoundExpression? result)
+        {
+            var preparedArgs = ArrayBuilder<BoundExpression>.GetInstance(capacity: args.Count);
+
+            var needsSpanRefParamConstructor = false;
+            var needsImplicitConversionFromStringToSpan = false;
+
+            NamedTypeSymbol? charType = null;
+
+            foreach (var arg in args)
+            {
+                Debug.Assert(arg.HasAnyErrors || arg.Type?.IsStringType() == true);
+
+                if (arg is BoundCall { ReceiverOpt: { Type: NamedTypeSymbol { SpecialType: SpecialType.System_Char } receiverCharType } receiver } potentialToStringCall &&
+                    (object)potentialToStringCall.Method != _compilation.GetSpecialTypeMember(SpecialMember.System_Object__ToString) &&
+                    (object)potentialToStringCall.Method.GetLeastOverriddenMethod(charType) == _compilation.GetSpecialTypeMember(SpecialMember.System_Object__ToString))
+                {
+                    needsSpanRefParamConstructor = true;
+                    charType = receiverCharType;
+                    preparedArgs.Add(receiver);
+                    continue;
+                }
+
+                preparedArgs.Add(arg);
+                needsImplicitConversionFromStringToSpan = true;
+            }
+
+            WellKnownMember? concatMember = preparedArgs.Count switch
+            {
+                2 => WellKnownMember.System_String__Concat_2ReadOnlySpans,
+                3 => WellKnownMember.System_String__Concat_3ReadOnlySpans,
+                4 => WellKnownMember.System_String__Concat_4ReadOnlySpans,
+                _ => null,
```

Comment:
>null [](http://example.com/codeflow?start=21&length=4)

Is this branch reachable?

Reply:
Considering https://github.com/dotnet/roslyn/pull/71793#discussion_r1479041298, it will be 100% not. But I have to provide some default case. Is it ok to throw an "unreachable" exception or it is too much risk for a compiler codebase?

Reply:
> . Is it ok to throw an "unreachable" exception or it is too much risk for a compiler codebase?

This is actually what we do in cases like this.